### ISSUE TITLE
Implement watchdog support

### ISSUE THREAD

**Issue by ants** [zalando/patroni#239]
When Patroni dies or wedges itself split brain protection is not guaranteed. Adding watchdog support would give reasonably good guarantees of PostgreSQL master going away when Patroni dies. Watchdog driver will ensure that the server gets rebooted when the heartbeat stops.

For now I'm researching two implementation avenues:
1. Create check script and example config for [watchdog daemon](http://www.sat.dundee.ac.uk/psc/watchdog/Linux-Watchdog.html).
2. Use [Linux kernel watchdog API](https://www.kernel.org/doc/Documentation/watchdog/watchdog-api.txt) directly.

The daemon option is probably more fully featured and integrates better with other liveness checks. Using the kernel API directly is simpler, avoids an external dependency and integration with Patroni lifecycle is simpler (initialize watchdog when Patroni starts, ping watchdog once per loop, close watchdog when gracefully exiting).

Is this wanted? Any preference for which way to go about this?

Patch coming up soon.

**Comment by CyberDem0n** [comment]
> For now I'm researching two implementation avenues:
>    1. Create check script and example config for watchdog daemon.
>    2. Use Linux kernel watchdog API directly.

Every option has it's own props and cons, for example:
 * calling of external script on the loaded system could work not very reliable. Sometimes it could take seconds to execute a program, which might cause a reboot. 
 * from another side, implementing `Linux kernel watchdog API` will make Patroni Linux dependent.
 * although this feature probably must not be enabled by default and it should be relatively easy to implement auto-detection of operating system.

> Is this wanted? Any preference for which way to go about this?

Personally we don't need this, but if you came up with such idea,  it could be that somebody else also needs it. So... why not.

**Comment by ants** [comment]
I do have a client that wants this additional layer of split brain protection. I have deployed watchdog and SBD based clusters with Pacemaker and actually found it much more robust than STONITH based approaches.

I'm currently proceeding with the integrated approach. The main benefits of the external watchdog don't matter much in my opinion:

* The external watchdog can monitor various system parameters like running out of memory or too high load. In my experience the cause for such issues is more often than not system independent, in which case doing a hard failover only makes things worse.
* The external watchdog can try to repair the system by e.g. restarting Patroni. But I'm not sure how many failovers will this prevent. I think doing our best to not crash, and rebooting the host when we do is sufficient.

I'm coding the watchdog module to support platform dependent implementations. For now I'm only doing Linux, but BSD support would look pretty much the same, only with different ioctl values for the operations. I'm sure Windows and other platforms could also be supported if someone really has enough interest to look up the documentation for it.

I'm kind of undecided whether the default for the feature should be off or automatic. Automatic meaning that it will try to use the watchdog interface if it's available and it has permissions for it. In that case enabling watchdog support would be as easy as `modprobe softdog && chown postgres /dev/watchdog`.

I will also make the watchdog support cluster aware. I want the watchdog to be disabled when only one instance remains in the cluster. If you do run into a situation where watchdog triggers due to high load the last thing you want is to kill the last remaining instance. Probably will also want to add an exponential backoff to the TTL timeout to avoid failover ping-pong failure mode. But this looks like a separate feature.

Basically the idea is to make Patroni not do too much damage in cases where competent DBA response time can be multiple hours.

**Comment by ainlolcat** [comment]
Will this solution be compatible with docker/kubernetes/openshift? They implement their own watchdogs (check if process alive or can check http response) and it would be issue if docker container will fail without module. Also it wolud be issue if module will try to restart main process (if enabled by default).

**Comment by ants** [comment]
I am not familiar with docker/kubernets/openshift watchdog. By the sound of it that mechanism can not ensure timely termination of PostgreSQL if Patroni does not get to perform demotion quickly enough. Does it even kill PostgreSQL if Patroni dies?

If a containerization or virtualization platform wants to do watchdog support it should have a way to terminate the container/VM if it doesn't check in often enough. And for better reliability the container manager/hypervisor should have its own watchdog that kills the whole host when container isolation is not completed in a timely manner. Xen does provide a watchdog driver, which would be compatible with this, but I haven't found any details about what it actually does. However I think that the softdog kernel driver approach is reliable enough even when running in a VM.

**Comment by CyberDem0n** [comment]
If the Patroni running in a docker with pid = 1 and it (Patroni) died - the entire container will die, together with postgres.

If your platform is able to run health-checks via http - it is fine, Patroni has already REST API.
```console
GET /patroni -- will return you some basic node information with response code 200
GET /master -- will return you 200 only if the node has the lock and postgres is not running in recovery
GET /replica -- will return you 200 if the node is running as a replica (it does not check replication delay though...)
```

**Comment by ainlolcat** [comment]
As @CyberDem0n pointed out docker already has reliable watchdog for root process and it can be enhanced with kubernetes/openshift livenessProbe (http check for example). I just worried if new watchdog can cause problem with dockerization. For example there is no root privileges (by default) and You cannot kill init process (which patroni will be in general). Even `chown ` command is disabled. So if patroni will try to execute `modprobe softdog && chown postgres /dev/watchdog` during start inside container it will crash. It would be nice to have "disable" button at least.

**Comment by ants** [comment]
A simple liveness probe is not enough for what I am implementing. It catches the most likely problem (Patroni crashes), but it doesn't do anything about other failure modes. For example if HA loop gets wedged due to a bug in Patroni, the container does not get a time slice for a while due to higher priority activity on the host, somebody suspends the VM for a while, etc. That said, the API thread could provide an interface to tell the external liveness probe if HA loop has been running in a timely manner. This would provide some additional robustness, depending on how timely the liveness checks and restarts are (Kubernetes documentation is somewhat sparse on those points).

Patroni will not execute either `modprobe` or `chown`, these are sysadmins responsibility. I have a systemd oneshot unit that does that before Patroni service is started. It's not yet decided if the default will be disabled or automatic. The automatic mode will just see if Patroni has permissions to open `/dev/watchdog`. If not, Patroni works just like it does now.

The watchdog support boils down to strategically placed calls in HA loop that tell an external agent "this node is safe from causing split brain for another ttl/2 seconds." The watchdogs job is to ensure that if that time ever runs out then anything that is still working on the node will not have a chance to send out a successful commit reply.

**Comment by CyberDem0n** [comment]
I think we can close it, https://github.com/zalando/patroni/pull/343 and https://github.com/zalando/patroni/pull/452 already solving a task.
If further development is needed we can open a separate issue.
